### DISCUSSION TITLE
Iperf3: how to use SCTP with several interfaces

### DISCUSSION THREAD

**Discussion by Neestoor** [esnet/iperf#1320] [Q&A]
Hi, I am trying SCTP with iperf3. I have checked two different options, with 
-P and --nstreams. In this case I had only one interface. First I used “iperf3 
– c IPsrv -t 4 -i 2 –SCTP  -P 2” to parallel several streams options with 
results around 600Mbps for each connection. Then, I used “iperf3 – c IPsrv 
-t 4 -i 2 –SCTP  -nstreams 2” but in this case the results were very bad, 
around 260 kbps and 0 kbps respectively. Do you know what it is going on?

In addition, I would like to try SCTP using simultaneously different 
interfaces at the same host, with different IP address. How can I call iperf3 in order to 
use these interfaces with SCTP under only one SCTP association? Do you have any 
example?

Thanks in advance.

root@ubuntu:/home/cliente/Downloads# iperf3 -c 192.168.130.132 -t 4 -i 2 --sctp --nstreams 2
Connecting to host 192.168.130.132, port 5201
[  5] local 192.168.130.131 port 45761 connected to 192.168.130.132 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-2.00   sec  64.0 KBytes   262 Kbits/sec                  
[  5]   2.00-4.00   sec  0.00 Bytes  0.00 bits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-4.00   sec  64.0 KBytes   131 Kbits/sec                  sender
[  5]   0.00-4.00   sec  64.0 KBytes   131 Kbits/sec                  receiver

iperf Done.
root@ubuntu:/home/cliente/Downloads# iperf3 -c 192.168.130.132 -t 4 -i 2 --sctp -P 2
Connecting to host 192.168.130.132, port 5201
[  5] local 192.168.130.131 port 45226 connected to 192.168.130.132 port 5201
[  7] local 192.168.130.131 port 57767 connected to 192.168.130.132 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-2.01   sec   167 MBytes   697 Mbits/sec                  
[  7]   0.00-2.01   sec   164 MBytes   684 Mbits/sec                  
[SUM]   0.00-2.01   sec   331 MBytes  1.38 Gbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[  5]   2.01-4.00   sec   169 MBytes   711 Mbits/sec                  
[  7]   2.01-4.00   sec   168 MBytes   708 Mbits/sec                  
[SUM]   2.01-4.00   sec   337 MBytes  1.42 Gbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-4.00   sec   336 MBytes   704 Mbits/sec                  sender
[  5]   0.00-4.00   sec   336 MBytes   704 Mbits/sec                  receiver
[  7]   0.00-4.00   sec   332 MBytes   696 Mbits/sec                  sender
[  7]   0.00-4.00   sec   332 MBytes   696 Mbits/sec                  receiver
[SUM]   0.00-4.00   sec   668 MBytes  1.40 Gbits/sec                  sender
[SUM]   0.00-4.00   sec   668 MBytes  1.40 Gbits/sec                  receiver

iperf Done.

**Comment by bmah888** [comment]
I don't think iperf3 supports the multipathing functionality you're asking about. The SCTP support is pretty rudimentary and I don't think it gets used very much in general (we don't use this within ESnet at all). It's also possible that it's buggy and/or not-well-performing, which could explain the performance values you saw. (Another thought is that the kernel SCTP implementation might not be as well-performing as TCP.) Wish I had a better answer for you, but I think fixing these problems would take some investigation and/or development effort, probably by somebody other than ESnet.

**Comment by davidBar-On** [comment]
Regarding the specific issues raised (I can't help with the general issue of SCTP support):
1. Did you check the server output when `--nstreams 2` was set?  My guess is that the 64K is the buffer size of the client machine, and that no message was received by the server.  If this is the case, maybe there is a problem in the SCTP settings in the client and/or server machines.
2. Regarding "using simultaneously different interfaces at the same host" - can the `--xbind` option help?  Note that this option can repeat several times in the command line.

**Comment by BUKIdrona** [comment]
Hi, I see that there were issues with --nstreams in SCTP mode. Has this been resolved in any recent updates to iperf3? If not, are there any known workarounds? Thanks!
